ff-fft/src/domain/radix2.rs

```diff
@@ -0,0 +1,312 @@
+//! This module defines `Radix2EvaluationDomain`, an `EvaluationDomain`
+//! for performing various kinds of polynomial arithmetic on top of
+//! fields that are FFT-friendly. `Radix2EvaluationDomain` supports
+//! FFTs of size at most `2^F::TWO_ADICITY`.
+
+pub use crate::domain::utils::Elements;
+use crate::domain::{
+    utils::{best_fft, bitreverse},
+    DomainCoeff, EvaluationDomain,
+};
+use crate::Vec;
+use algebra_core::{FftField, FftParameters};
+use core::fmt;
+#[cfg(feature = "parallel")]
+use rayon::prelude::*;
+
+/// Defines a domain over which finite field (I)FFTs can be performed. Works
+/// only for fields that have a large multiplicative subgroup of size that is
+/// a power-of-2.
+#[derive(Copy, Clone, Hash, Eq, PartialEq)]
+pub struct Radix2EvaluationDomain<F: FftField> {
+    /// The size of the domain.
+    pub size: u64,
+    /// `log_2(self.size)`.
+    pub log_size_of_group: u32,
+    /// Size of the domain as a field element.
+    pub size_as_field_element: F,
+    /// Inverse of the size in the field.
+    pub size_inv: F,
+    /// A generator of the subgroup.
+    pub group_gen: F,
+    /// Inverse of the generator of the subgroup.
+    pub group_gen_inv: F,
+    /// Multiplicative generator of the finite field.
+    pub generator_inv: F,
+}
+
+impl<F: FftField> fmt::Debug for Radix2EvaluationDomain<F> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Radix-2 multiplicative subgroup of size {}", self.size)
+    }
+}
+
+impl<F: FftField> EvaluationDomain<F> for Radix2EvaluationDomain<F> {
+    type Elements = Elements<F>;
+
+    /// Construct a domain that is large enough for evaluations of a polynomial
+    /// having `num_coeffs` coefficients.
+    fn new(num_coeffs: usize) -> Option<Self> {
+        // Compute the size of our evaluation domain
+        let size = num_coeffs.next_power_of_two() as u64;
+        let log_size_of_group = size.trailing_zeros();
+
+        // libfqfft uses > https://github.com/scipr-lab/libfqfft/blob/e0183b2cef7d4c5deb21a6eaf3fe3b586d738fe0/libfqfft/evaluation_domain/domains/basic_radix2_domain.tcc#L33
+        if log_size_of_group > F::FftParams::TWO_ADICITY {
+            return None;
+        }
+
+        // Compute the generator for the multiplicative subgroup.
+        // It should be 2^(log_size_of_group) root of unity.
+        let group_gen = F::get_root_of_unity(size as usize)?;
```

Comment:
Maybe we want to `debug_assert!` that this is indeed a root of unity for the given domain? (Also maybe for  `MixedRadixEvaluationDomain`)

Reply:
I added it as well as actual tests for it.